[RFC] Colour quoted text

Nelyah opened the pull request:
## Context 

I feel like being able to render quote text according to the quoted level with colours is important. At least it is for me in terms of readability.

This PR is to answer this particular feature. An issue has been opened #746, but it was answered with #612 and then #1015. These are great features, but don't quite target this specific issue.  

## Modifications

I modified the file `alot/widget/thread.py` to call the `parse_text_colour` function, if we are looking at words linewise. The function `parse_text_colour` asks the config to check if we want to parse quoted text. If yes, it proceeds and calls the function `parse_quotes`. 

The latter returns the theme value associated with the quote_level in the theme file (if there is one). 

## Example configuration

This is working well for me and it passes the tests. This is how I set it up in my config file to make either `>` or four spaces a quote symbol.:

    quote_symbol = '( {4}|>)'
    parse_quotes = True

I am fairly new to this project, please let me know if there are obvious things I am missing.

Nelyah commented:
Thank you @pazz, this is a wonderful piece of software you got there!

Here is a screenshot of a sample mail. I set up the colours being close from one another, but this can be changed with the options `quote_level_i` in the configuration file (where i is the quotation level wished for). I declared those only up to i=7. Let me know if you'd wish more colour contrast.

![Screenshot from 2019-10-10 18-17-58](https://user-images.githubusercontent.com/6171842/66610506-01e9f680-ebbc-11e9-82f7-7788ec738471.png)


> Would this only change the urwid representation or does your code changes the text of follow-up reply emails?

No, this should only change the urwid representation. The only real modification to existing code is to change the `attr` theming value (in `alot/widget/thread.py`). I am getting the new `attr` value by parsing the mail body and searching it for quotation markers (which can be defined in the configuration file)

lucc commented on alot/widgets/colour_text.py:
```diff
@@ -0,0 +1,55 @@
+    quote_colour = None
+
+    for quote_level in range(1, max_quote_level+1):
+        quote_regex = r'^\ *(' + symbol + '\ *){' + str(quote_level) + r'}'
```
Why are there backslashes before whitespace? Why are these raw strings? Please use `str.format` instead of `str+str`.

Nelyah replied:
You're right, I pushed a commit to use `str.format` instead of `str+str`.

As far as I know however, it is good practice to use raw strings when writing regexes (which may contain special characters). It is not yet the case in this one but I prefer to do it consistently.

I escape the space characters to make them more visible and improve the regex readability. Should I delete them? They are not otherwise necessary.

lucc replied:
I think escaping stuff that doesn't need escaping does not make the code more readable. It just makes me wonder weather I or the author did not understand (the) regex(es) .

Nelyah replied:
Fair enough. I fixed it in the last push.

lucc commented on alot/widgets/colour_text.py:
```diff
@@ -0,0 +1,55 @@
+        return parse_quotes(line)
+
+
+def parse_quotes(line, max_quote_level=7):
```
Maybe it is better to move the 7 into a local variable. The argument is not changed from anywhere (if I see that correctly). Please also document that there is the magic number 7 here because there are 7 levels predefined in the config spec. I think the code would work if a user defines a higher level in their config that is not in the spec. But I don't think we want to support this case.

Please fix the newline in the docstring. It should be between the text and the parameters not the params and the return.

Nelyah replied:
I pushed another commit to move the 7 to a local variable. I explain in the docstring up to what level we go and in a comment why is that.

From what I have tested, if a user defines `quote_level_8` (`>7`) in their theming config, that attribute is not considered. When a line is quoted, my code will match the highest possible quote level up to 7, then apply the theming attribute of that highest quote level found.

For instance:

    This is not a quoted line
    >>>>>>> This is 7th quote level
    >>>>>>>> This is 8th quote level
    >>>>>>>>> This is 9th quote level

Except for the not quoted line, they will all match the theming attribute of the 7th quote level. If no such attribute was defined in the theming config or if the line is not quoted, then the return value of `settings.get_theming_attribute` is `'default'`. If such a value is returned, we use the `theme.body` theming attribute instead.

lucc commented on alot/widgets/colour_text.py:
```diff
@@ -0,0 +1,55 @@
+        return None
+
+
+def get_quote_colour(line, max_quote_level):
```
It would be good to provide some tests for this code.

Nelyah replied:
Ok! I'll work on it.

Nelyah commented:
I just added the tests as well

lucc commented on tests/widgets/test_colour_text.py:
```diff
@@ -0,0 +1,110 @@
+        'alot.commands.globals.settings._theme',
+        theme.Theme(DUMMY_THEME.splitlines()))
+    def test_quote_level_get_equals_corresponding_attribute(self):
+        self.assertEqual(settings.get('quote_symbol'), '>')
```
We don't have to test `settings.get` here.

Nelyah replied:
Fixed in the last push.

lucc commented on tests/widgets/test_colour_text.py:
```diff
@@ -0,0 +1,110 @@
+        focus = '', '', '', '', '', ''
+        normal = '', '', '', '', '', ''
+[thread]
+    quote_level_2 = '', '', '', '', '#880', '#886'
```
Please use different colors for each value so that no false positive can happen because of color code collision.  I think best would be "recognizable" codes: `#201` and `#202` for level 2, `#301` and `#302` for level 3 ...

Nelyah replied:
Fixed in the last push. I also moved the theme variable as global to my script as to avoid reading it for each test case.

lucc commented on tests/widgets/test_colour_text.py:
```diff
@@ -0,0 +1,110 @@
+    def test_quote_level_get_equals_corresponding_attribute(self):
+        self.assertEqual(settings.get('quote_symbol'), '>')
+
+        self.assertIsNone(parse_quotes(''))
```
I would put these None cases in one test method of their own.

Nelyah replied:
Done. I put together default values for quotes and non quotes in the same method as they return `None`.

lucc commented on tests/widgets/test_colour_text.py:
```diff
@@ -0,0 +1,110 @@
+        self.assertIsNone(parse_quotes('test no quote'))
+
+        # Undefined value
+        self.assertIsNone(parse_quotes('>'))
```
Why is this None?  I was expecting it to be the color for level 1. But that is not defined. So maybe this is the default value from the spec?  In this case I would make one extra test method that tests default values.

Nelyah replied:
When an undefined `quote_level` theming attribute is requested but not found, `'default'` is returned by `settings.get_theming_attribute`.

The return value from `get_quote_colour` is either `None` if no quote were found, `'default'` if a quote was found but the theming was undefined or an `AttrSpec` instance if this is a quote but its theming value is defined.

In the first two cases, `parse_quotes` will return `None` and this is wanted.   
>  In this case I would make one extra test method that tests default values.

I added on such test case.

lucc commented on tests/widgets/test_colour_text.py:
```diff
@@ -0,0 +1,110 @@
+    def test_regex_symbol_is_correct(self):
+        # Match 4th level quote
+        self.assertEqual(parse_quotes('Aaa A'), AttrSpec('#580', '#586'))
+        self.assertEqual(parse_quotes('Aa<>tb A'), AttrSpec('#580', '#586'))
```
I would also add tests for stuff like `some line with >> marker in the middle`, `some line with the marker at the end: >>` .

Nelyah replied:
Done.

Nelyah commented:
Thank you for the feedback, @lucc! And sorry if this has taken some time and back and forth.

pazz commented on alot/widgets/colour_text.py:
```diff
@@ -0,0 +1,42 @@
+import logging
```
please add a licence header as in all other files. You are welcome to add your name in there of course

Nelyah replied:
Just added it

pazz commented on alot/widgets/colour_text.py:
```diff
@@ -0,0 +1,42 @@
+
+        if re.match(quote_regex, line):
+            is_quoted = True
+            logging.info(f'Requesting attribute quote_level_{quote_level}')
```
should be `debug`, not `info` imo

pazz commented on alot/widgets/colour_text.py:
```diff
@@ -0,0 +1,42 @@
+    is_quoted = False
+    quote_colour = None
+    symbol = settings.get('quote_symbol')
+    if symbol is None:
```
use the fallback option when getting the config setting? https://github.com/pazz/alot/blob/master/alot/settings/manager.py#L239
(note that the defaul is "> ", with a trailing space)

Nelyah replied:
Does it  make sense here to specify a fallback? If I understand correctly we would still get the default settings value of the variable we're requesting (in this case from `quote_symbol` : `>`), even if no fallback is defined.

pazz commented on alot/widgets/colour_text.py:
```diff
@@ -9,7 +9,8 @@ def parse_text_colour(line):
 
     """
-    return parse_quotes(line)
+    if settings.get('parse_quotes'):
```
otherwise return `None` (implicitly)?

pazz commented on alot/defaults/alot.rc.spec:
```diff
@@ -17,6 +17,10 @@ input_timeout = float(default=1.0)
 
+# Show quotes be parsed
+parse_quotes = boolean(default=True)
+quote_symbol = string(default='>')
```
why not use the existing `quote_prefix`?

Nelyah replied:
I didn't see it before!  

However, I notice that it has a space added to it. That would mean that a line like:

```
>> Level 2 quote level
```

wouldn't get interpreted correctly, because the regex `r'^ *({} *){{{}}}'.format(symbol, quote_level)` wouldn't match. It would still be possible to change it but the default wouldn't work really well. 

I'm not sure in what context  `quote_prefix` is used. A first solution would be to remove the space and accommodate for this wherever it is used. A second could be to rename my `quote_symbol` to something more explicit, like `quotation_regex_detection`.

lenormf replied:
I don't think this new option is necessary. The two most commonly used characters for quoting text are `>` and `|`, and I've only ever seen the bracket being used ([wikipedia reference](https://en.wikipedia.org/wiki/Posting_style#Quoted_line_prefix)).

It would make the PR leaner to assume only those two characters highlight quotes in messages, plus it would avoid some weird unit tests verifying that the code behaves as expected when using `[Aa]` for `quote_symbol`.

Nelyah commented:
I'll do the licence change and add the quote_prefix (didn't see it previously). 
However squashing commits (which I have already pushed to origin) through a rebase means I am rewriting all of my commits as my new master diverges from origin/master, isn't it? 
I'm not quite a git professional, but let me know if this is what I should do (unless there is a better solution for it)

lucc commented:
You can do an interactive rebase. Like this `git rebase --interactive origin/master message-body-colour`. Learn more about it [from the docs](https://help.github.com/en/github/using-git/about-git-rebase).

pazz commented:
Where are we on this? It'd be a shame to let this get too old and diverge from master too much to be useful..

Nelyah commented:
Oh, I'm sorry. 

I had some things happening at the time and it just slipped out of my mind. I will get back to this in the next few days! Sorry again for the delay...

Nelyah commented:
I rebased and resolved a few merge conflicts (mainly in `thread.py` with `FocusableText` being replaced by `ANSIText`), but have not yet added anything else.

I added a few `[tags]` to my commits to help myself (and maybe you as well) since it's been a while I worked on this. I'll remove them before merging this.

pazz commented on tests/widgets/test_colour_text.py:
```diff
@@ -0,0 +1,119 @@
+from urwid import AttrSpec
+from unittest import mock
+
+from alot.settings.const import settings
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused settings imported from alot.settings.const](https://app.codacy.com/manual/patricktotzke/alot/pullRequest?prid=4429219)

Nelyah replied:
Removed it in 32cebe8

lenormf commented on alot/widgets/colour_text.py:
```diff
@@ -0,0 +1,65 @@
+    symbol = settings.get('quote_symbol')
+    quote_colour = None
+
+    for quote_level in range(1, max_quote_level+1):
```
I think running 7 regex matches on a line that has 7 levels of nested quoting is too much.

I made it work (with a dirty patch) on my local branch with only two calls to `re.match()`, for any amount of nesting, I think it's the better way to go about it.

Plus if we remove the `quote_symbol` option (c.f. my other comment), it will be easier to parse spaces properly ([RFC reference](https://tools.ietf.org/html/rfc3676#section-4.5))

lenormf commented on tests/widgets/test_colour_text.py:
```diff
@@ -0,0 +1,118 @@
+            'colourmode': 256
+            }
+
+SETTINGS_test_symbol_regex = {'quote_symbol': '([Aa]|<>.b|>)',
```
As mentioned in another comment, this is a weird thing to test for. Maybe we don't need the code to be that generic.

lenormf commented on alot/widgets/colour_text.py:
```diff
@@ -0,0 +1,65 @@
+
+        else:
+            # If there is no match at some point,
+            # we simply use the last level match colour
```
I think cycling back to the colours for the first level of quoting is a better option:

* 8 → 1
* 9 → 2
* …
* 15 → 1
* 16 → 2

etc.

Basically a modulo.
